src/coreclr/jit/morphblock.cpp

```diff
@@ -1154,25 +1157,86 @@ GenTree* MorphCopyBlockHelper::CopyFieldByField()
+            addr = m_store->AsIndir()->Addr();
+
+            m_comp->gtPeelOffsets(&addr, &addrBaseOffs, &addrBaseOffsFldSeq);
 
-            // "dstAddr" might be a complex expression that we need to clone
+            // "addr" might be a complex expression that we need to clone
             // and spill, unless we only end up using the address once.
             if (m_srcVarDsc->lvFieldCnt > 1)
             {
-                if (CanReuseAddressForDecomposedStore(dstAddr))
+                if (CanReuseAddressForDecomposedStore(addr))
                 {
                     // "dstAddr" is simple expression. No need to spill
-                    noway_assert((dstAddr->gtFlags & GTF_PERSISTENT_SIDE_EFFECTS) == 0);
+                    noway_assert((addr->gtFlags & GTF_PERSISTENT_SIDE_EFFECTS) == 0);
                 }
                 else
                 {
-                    addrSpill = dstAddr;
+                    addrSpill = addr;
                 }
             }
         }
     }
 
+    if (addrSpill != nullptr)
+    {
+        // 'addrSpill' is already morphed
+
+        // Spill the (complex) address to a temp.
+        // Note, at most one address may need to be spilled.
+        addrSpillTemp = m_comp->lvaGrabTemp(true DEBUGARG("BlockOp address local"));
+
+        LclVarDsc* addrSpillDsc = m_comp->lvaGetDesc(addrSpillTemp);
+        addrSpillDsc->lvType = addrSpill->TypeIs(TYP_REF) ? TYP_REF : TYP_BYREF; // TODO-ASG: zero-diff quirk, delete.
+        addrSpillStore       = m_comp->gtNewTempAssign(addrSpillTemp, addrSpill);
+    }
+
+    auto grabAddr = [=, &result](unsigned offs) {
+        GenTree* addrClone = nullptr;
+        // Need address of the source.
+        if (addrSpill)
+        {
+            assert(addrSpillTemp != BAD_VAR_NUM);
+            addrClone = m_comp->gtNewLclvNode(addrSpillTemp, addrSpill->TypeGet());
+        }
+        else
+        {
+            if (result == nullptr)
+            {
+                // Reuse the original address tree for the first field.
+                addrClone = addr;
+            }
+            else
+            {
+                // We can't clone multiple copies of a tree with persistent side effects
+                noway_assert((addr->gtFlags & GTF_PERSISTENT_SIDE_EFFECTS) == 0);
+
+                addrClone = m_comp->gtCloneExpr(addr);
+                noway_assert(addrClone != nullptr);
+
+                JITDUMP("Multiple Fields Clone created:\n");
+                DISPTREE(addrClone);
+
+                // Morph the newly created tree
+                addrClone = m_comp->fgMorphTree(addrClone);
+            }
+        }
+
+        assert(addrClone != nullptr);
+        ssize_t fullOffs = addrBaseOffs + (ssize_t)offs;
+        if ((fullOffs != 0) || (addrBaseOffsFldSeq != nullptr))
+        {
+            // Avoid using unsigned overload of gtNewIconNode
+            // that takes field seq to get correct overflow
+            // handling.
```

Comment:
Speaking of overflow... There is a sense in which doing this arithmetic in `ssize_t`s could be incorrect on 64 bit hosts targeting 32 bit targets, as it could create out-of-range `INT` constants. I would expect this to not be a problem in the simple cases, however.

Reply:
I can add an `int32_t` truncation in that case to be safe.

Reply:
Or I guess the right fix is to switch `gtPeelOffsets` to `target_ssize_t`